spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,292 @@
+import org.apache.spark.sql.sources.v2.reader.streaming.MicroBatchReader;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A mirco-batch based Spark Structured Streaming reader for Iceberg table. It will track the added
+ * files and generate tasks per batch to process newly added files. By default it will process
+ * all the newly added files to the current snapshot in each batch, user could also set this
+ * configuration "max-files-per-trigger" to control the number of files processed per batch.
+ */
+class StreamingReader extends Reader implements MicroBatchReader {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReader.class);
+  private static final int DEFAULT_MAX_FILES_PER_TRIGGER = 1000;
+  private static final long CACHE_SIZE = 1000;
+
+  private Offset startOffset;
+  private Offset endOffset;
+
+  private final Table table;
+  private final int maxFilesPerTrigger;
+  private final long splitSize;
+  private final int lookback;
+  private final long openFileCost;
+  private final Offset initialOffset;
+
+  private final Cache<Long, CloseableIterable<FileScanTask>> snapshotTasksCache = Caffeine.newBuilder()
+      .maximumSize(CACHE_SIZE)
+      .softValues()
+      .build();
+
+  /**
+   * Utility class to track the snapshotId, task and index of task within this snapshot.
+   */
+  @VisibleForTesting
+  static class IndexedTask {
+    private final long snapshotId;
+    private final FileScanTask task;
+    private final int index;
+
+    IndexedTask(long snapshotId, FileScanTask task, int index) {
+      this.snapshotId = snapshotId;
+      this.task = task;
+      this.index = index;
+    }
+
+    long snapshotId() {
+      return snapshotId;
+    }
+
+    FileScanTask task() {
+      return task;
+    }
+
+    int index() {
+      return index;
+    }
+  }
+
+  StreamingReader(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryptionManager,
+                  boolean caseSensitive, DataSourceOptions options) {
+    super(table, io, encryptionManager, caseSensitive, options);
+
+    this.table = table;
+    this.maxFilesPerTrigger =
+      options.get("max-files-per-trigger").map(Integer::parseInt).orElse(DEFAULT_MAX_FILES_PER_TRIGGER);
+    Preconditions.checkArgument(maxFilesPerTrigger > 0,
+        "Option max-files-per-trigger '%d' should > 0", maxFilesPerTrigger);
+
+    Long snapshotId = options.get("snapshot-id").map(Long::parseLong).orElse(null);
+    if (snapshotId != null) {
+      if (SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), snapshotId)) {
+        this.initialOffset = new StreamingOffset(snapshotId, -1, true);
+      } else {
+        LOG.warn("The option snapshot-id {} is not a valid snapshot id, will use the earliest " +
+            "snapshot instead", snapshotId);
+        this.initialOffset = calculateInitialOffset();
```

Comment:
I think this should be an error instead of using the initial offset. The user has expressed a snapshot ID to process from and using the initial offset would not return the rows that the user requested. I think not being able to satisfy the user's query should be an exception instead of a warning. Users don't look at warnings, so this will cause silently incorrect results.

In addition, the initial offset is a start offset, so all existing data is processed. That could easily cause a failure because this will try to load all data files into memory at the same time.

Reply:
In the Kafka, if the specified offset is not existed, Kafka will automatically change to use earliest or latest offset based on the configuration. I was thinking to achieve the same functionality here. I can change to throw an exception.

 > In addition, the initial offset is a start offset, so all existing data is processed. That could easily cause a failure because this will try to load all data files into memory at the same time.

I'm not sure why this will cause an issue to load all data files into memory. If we specify a starting offset, then current implementation will retrieve the complete data from this snapshot plus all the following snapshots' appended files. This has no big difference compared to travel from very beginning of the table. Would you please help to explain more?

Reply:
The case I'm thinking about is when `is_starting` is true, so the source will try to read all existing data in the table for the first known snapshot. That will cause `getChanges` to build [a scan of all data files](https://github.com/apache/incubator-iceberg/pull/796/files#diff-ad2e27f21c81a539b459360e10844039R249-R250), which is then [sorted and collected into a list](https://github.com/apache/incubator-iceberg/pull/796/files#diff-ad2e27f21c81a539b459360e10844039R287-R289).

Reply:
To solve this problem, what about changing how we get a reliable ordering for the data files? Because each snapshot is immutable, the data files for a snapshot are always stored across the same set of manifest files. Instead of sorting data files, you could sort the manifest files in a snapshot. Then you'd be able to use `ManifestGroup` to reliably build an iterable of `FileScanTask` in a reliable order, without sorting a large list.

That would decrease planning time as well as memory use because this wouldn't need to load all of the data files at one time.

Reply:
@rdblue , I'm not sure if I understand correctly. If we sort manifest files in a snapshot, then can we guarantee that the `FileScanTask` is stable sorted when generating using parallel thread pool?

Reply:
I get your point now, let me try to change it.

Reply:
I think the way of using manifest files requires to change several parts of Iceberg core to expose the logics of plan files, not sure if this is an elegant way.

1. How do we control the process per batch, now the semantic changes to "manifests per trigger", there could bunch of data files in one manifest, how do we semantically control the size per batch.
2. The major problem problem mentioned above could only happen when reader is far behind writer. If the reader could catch up with writer, then in each batch it will not process two many snapshots, and the plan time should not be long.

So I'm still thinking of this way to use manifest files and expose manifest group.

Reply:
I would defer the changes to use manifest group, need to figure out a clean way.